src/main/kotlin/org/rust/lang/refactoring/extractFunction/RsExtractFunctionHandler.kt

```diff
@@ -0,0 +1,73 @@
+package org.rust.lang.refactoring.extractFunction
+
+import com.intellij.openapi.actionSystem.DataContext
+import com.intellij.openapi.application.Result
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiParserFacade
+import com.intellij.refactoring.RefactoringActionHandler
+import org.rust.ide.utils.findStatementsInRange
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.parentOfType
+
+enum class RsWrapperType {
+    ImplMethod,
+    TraitMethod,
+    Function
+}
+
+class RsExtractFunctionHandler : RefactoringActionHandler {
+    override fun invoke(project: Project, elements: Array<out PsiElement>, dataContext: DataContext?) {
+        //this doesn't get called form the editor.
+    }
+
+    override fun invoke(project: Project, editor: Editor?, file: PsiFile?, dataContext: DataContext?) {
+        if (file !is RsFile) return
+        val start = editor?.selectionModel?.selectionStart
+        val end = editor?.selectionModel?.selectionEnd
+        if (start !== null && end !== null) {
+            val elements = findStatementsInRange(file, start, end).asList()
+            if (elements.isEmpty()) return
+            val first = elements.first()
+            val last = elements.last()
+
+            // check element should be a part of one block
+            val parentOfFirst = first.parentOfType<RsFunction>() ?: return
+            val parentOfLast = last.parentOfType<RsFunction>() ?: return
```

Comment:
Ideally, we should show some kind of error message here instead of silently doing nothing.